File: admission-system/backend/app/schemas/form_config.py
"""
Pydantic schemas for Form Configuration API
"""
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from datetime import datetime


# ============= Form Field Schemas =============

class FormFieldMasterResponse(BaseModel):
    id: int
    field_code: str
    field_label: str
    field_type: str
    category: str
    default_step: int
    is_required_by_default: bool
    validation_rules: Optional[Dict[str, Any]] = None
    help_text: Optional[str] = None
    placeholder: Optional[str] = None
    options: Optional[List[str]] = None
    file_config: Optional[Dict[str, Any]] = None
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


# ============= Form Configuration Schemas (Named Forms) =============

class FormConfigurationCreate(BaseModel):
    form_name: str
    form_description: Optional[str] = None
    is_active: bool = True
    template_id: Optional[int] = None  # Optional template to start from


class FormConfigurationUpdate(BaseModel):
    form_name: Optional[str] = None
    form_description: Optional[str] = None
    is_active: Optional[bool] = None


class FormConfigurationResponse(BaseModel):
    id: int
    school_id: int
    form_name: str
    form_description: Optional[str]
    is_active: bool
    created_by: Optional[int]
    created_at: datetime
    updated_at: Optional[datetime]
    field_count: int = 0  # Computed field

    class Config:
        from_attributes = True


class FormConfigurationDetailResponse(BaseModel):
    id: int
    school_id: int
    form_name: str
    form_description: Optional[str]
    is_active: bool
    created_by: Optional[int]
    created_at: datetime
    updated_at: Optional[datetime]
    field_configurations: List["SchoolFormConfigResponse"]

    class Config:
        from_attributes = True


# ============= School Form Configuration Schemas =============

class SchoolFormConfigCreate(BaseModel):
    field_id: int
    is_enabled: bool = True
    is_required: Optional[bool] = None
    step_number: int
    display_order: int
    conditional_logic: Optional[Dict[str, Any]] = None
    custom_label: Optional[str] = None
    custom_help_text: Optional[str] = None


class SchoolFormConfigUpdate(BaseModel):
    is_enabled: Optional[bool] = None
    is_required: Optional[bool] = None
    step_number: Optional[int] = None
    display_order: Optional[int] = None
    conditional_logic: Optional[Dict[str, Any]] = None
    custom_label: Optional[str] = None
    custom_help_text: Optional[str] = None


class SchoolFormConfigResponse(BaseModel):
    id: int
    form_config_id: int  # NEW - links to specific form
    school_id: int
    field_id: int
    is_enabled: bool
    is_required: Optional[bool]
    step_number: int
    display_order: int
    conditional_logic: Optional[Dict[str, Any]] = None
    custom_label: Optional[str] = None
    custom_help_text: Optional[str] = None
    created_at: datetime

    # Include field details
    field: FormFieldMasterResponse

    class Config:
        from_attributes = True


# ============= Template Schemas =============

class FormTemplateResponse(BaseModel):
    id: int
    template_name: str
    template_code: str
    description: Optional[str]
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class FormTemplateFieldResponse(BaseModel):
    id: int
    template_id: int
    field_id: int
    is_enabled: bool
    is_required: Optional[bool]
    step_number: int
    display_order: int
    field: FormFieldMasterResponse

    class Config:
        from_attributes = True


class FormTemplateDetailResponse(BaseModel):
    id: int
    template_name: str
    template_code: str
    description: Optional[str]
    is_active: bool
    created_at: datetime
    template_fields: List[FormTemplateFieldResponse]

    class Config:
        from_attributes = True


# ============= Apply Template Request =============

class ApplyTemplateRequest(BaseModel):
    template_id: int


# ============= School Configuration Summary =============

class SchoolFormConfigSummary(BaseModel):
    total_fields: int
    enabled_fields: int
    required_fields: int
    fields_by_step: Dict[int, int]  # {step_number: field_count}
    fields_by_category: Dict[str, int]  # {category: field_count}


# ============= Bulk Update =============

class BulkFieldUpdate(BaseModel):
    field_configs: List[SchoolFormConfigCreate]

File: admission-system/frontend/web-app/src/types/formConfig.ts
/**
 * TypeScript types for Form Configuration
 */

export interface FormFieldMaster {
  id: number;
  field_code: string;
  field_label: string;
  field_type: string;
  category: string;
  default_step: number;
  is_required_by_default: boolean;
  validation_rules?: Record<string, any>;
  help_text?: string;
  placeholder?: string;
  options?: string[];
  file_config?: Record<string, any>;
  is_active: boolean;
  created_at: string;
}

export interface FormConfiguration {
  id: number;
  school_id: number;
  form_name: string;
  form_description?: string;
  is_active: boolean;
  created_by?: number;
  created_at: string;
  updated_at?: string;
  field_count: number;
}

export interface FormConfigurationCreate {
  form_name: string;
  form_description?: string;
  is_active?: boolean;
  template_id?: number;
}

export interface FormConfigurationUpdate {
  form_name?: string;
  form_description?: string;
  is_active?: boolean;
}

export interface FormConfigurationDetail extends FormConfiguration {
  field_configurations: SchoolFormConfig[];
}

export interface SchoolFormConfig {
  id: number;
  form_config_id: number;
  school_id: number;
  field_id: number;
  is_enabled: boolean;
  is_required?: boolean;
  step_number: number;
  display_order: number;
  conditional_logic?: Record<string, any>;
  custom_label?: string;
  custom_help_text?: string;
  created_at: string;
  field: FormFieldMaster;
}

export interface FormTemplate {
  id: number;
  template_name: string;
  template_code: string;
  description?: string;
  is_active: boolean;
  created_at: string;
}

export interface FormTemplateField {
  id: number;
  template_id: number;
  field_id: number;
  is_enabled: boolean;
  is_required?: boolean;
  step_number: number;
  display_order: number;
  field: FormFieldMaster;
}

export interface FormTemplateDetail extends FormTemplate {
  template_fields: FormTemplateField[];
}

export interface SchoolFormConfigSummary {
  total_fields: number;
  enabled_fields: number;
  required_fields: number;
  fields_by_step: Record<number, number>;
  fields_by_category: Record<string, number>;
}

export interface FieldCategory {
  category: string;
  field_count: number;
}

export interface SchoolFormConfigCreate {
  field_id: number;
  is_enabled?: boolean;
  is_required?: boolean;
  step_number: number;
  display_order: number;
  conditional_logic?: Record<string, any>;
  custom_label?: string;
  custom_help_text?: string;
}

export interface SchoolFormConfigUpdate {
  is_enabled?: boolean;
  is_required?: boolean;
  step_number?: number;
  display_order?: number;
  conditional_logic?: Record<string, any>;
  custom_label?: string;
  custom_help_text?: string;
}

export interface ApplyTemplateRequest {
  template_id: number;
}

export interface BulkFieldUpdate {
  field_configs: SchoolFormConfigCreate[];
}
